### PR TITLE
Consider the requirement of a field to be readonly for record/object subtyping

### PR DESCRIPTION
**Pull request by MaryamZi**
## Purpose
$title - if the field in the target type is a `readonly` field, the relevant field in the source type should also be `readonly` either explicitly or due to the source type itself having its read-only bit set.

Fixes https://github.com/ballerina-platform/ballerina-lang/issues/23468

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java [hunk 1]
```diff
@@ -750,8 +750,16 @@ private boolean isAssignableMapType(BMapType sourceMapType, BRecordType targetRe
...
 
         for (BField field : targetRecType.fields.values()) {
-            if (!(Symbols.isFlagOn(field.symbol.flags, Flags.OPTIONAL) &&
-                    isAssignable(sourceMapType.constraint, field.type))) {
+            if (!Symbols.isFlagOn(field.symbol.flags, Flags.OPTIONAL)) {
+                return false;
+            }
+
+            if (Symbols.isFlagOn(field.symbol.flags, Flags.READONLY) &&
+                    !Symbols.isFlagOn(sourceMapType.flags, Flags.READONLY)) {
```
Comment (pubudu91) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java:758]: Should we consider extracting this out as a method? Seeing this repeated a couple of times + would probably be easier to read when used in more complex checks

**Comment by pubudu91** [comment]
Is the following disallowed by design?
```ballerina
type Person object {
    string name = "John";
    readonly int age;

    function __init(int age) {
        self.age = age;
    }
};
```
Gives the error
```
error: .::test.bal:13:9: cannot update 'readonly' object field 'age' in 'Person'
```

**Comment by MaryamZi** [comment]
> Is the following disallowed by design?
> 
> ```ballerina
> type Person object {
>     string name = "John";
>     readonly int age;
> 
>     function __init(int age) {
>         self.age = age;
>     }
> };
> ```
> 
> Gives the error
> 
> ```
> error: .::test.bal:13:9: cannot update 'readonly' object field 'age' in 'Person'
> ```

I assume you mean to set the value in the initializer method? If so it has to be `init` right?

The following doesn't give any errors.

```ballerina
type Person object {
    string name = "John";
    readonly int age;

    function init(int age) {
        self.age = age;
    }
};
```

The error in your sample is expected, because we don't allow setting values for object fields in methods other than in the `init`.

**Comment by pubudu91** [comment]
Argh.. Forgot we changed it to `init` :sweat_smile: Even in the `ini()`, should only be allowed to set it once right?
```ballerina
type Person object {
    string name = "John";
    readonly int age;

    function init(int age) {
        self.age = age;
        self.name = "foo";
        self.age = age + 20;
    }
};
```
This is probably not related to this PR though

**Comment by MaryamZi** [comment]
> Argh.. Forgot we changed it to `init`  Even in the `ini()`, should only be allowed to set it once right?
> 
> ```ballerina
> type Person object {
>     string name = "John";
>     readonly int age;
> 
>     function init(int age) {
>         self.age = age;
>         self.name = "foo";
>         self.age = age + 20;
>     }
> };
> ```
> 

Yeah, I believe the plan is to have this be an error. This is not yet implemented. https://github.com/ballerina-platform/ballerina-spec/issues/532 is the related spec issue and I just created https://github.com/ballerina-platform/ballerina-lang/issues/24119 to track fixing the implementation.

**Pull request merged**